### PR TITLE
Resolve some issues that mentioned in BEP031 pr changes document

### PR DESCRIPTION
**Pull request by konantian**
## Description
1. Set “BodyPartDetails” to be recommended
2. Restrict “BodyPartDetailsOntology” to be valid URI
3. Set the accepted values of “SampleEnvironement” to be “in vivo”, “ex vivo” or “in vitro”
4. “_photo” suffix extension MUST be either “.jpg”, “.png” or “.tif”
5. Change the data_type “microscopy” to “micr”

### CODE REVIEW

File: bids-validator/validators/json/schemas/microscopy.json [hunk 1]
```diff
@@ -62,12 +62,13 @@
     "InstitutionName",
     "InstitutionAddress",
     "InstitutionalDepartmentName",
-
     "BodyPart",
+    "BodyPartDetails",
     "Environment",
```
Comment (mariehbourget) [inline bids-validator/validators/json/schemas/microscopy.json]:
```suggestion
    "SampleEnvironment",
```
I just realized that `Environment` was changed to `SampleEnvironment` a while ago and we forgot to change it here.

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 2]
```diff
@@ -377,7 +377,9 @@
         "_channels\\.tsv",
         "_eeg\\.json",
         "_coordsystem\\.json",
-        "_photo\\.jpg"
+        "_photo\\.jpg",
+        "_photo\\.png",
+        "_photo\\.tif"
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
        "_photo\\.jpg"
```

The change in "photo" extensions only applies to "microscopy" modality, so not here for "eeg".
Sorry that was not clear in the list.

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 3]
```diff
@@ -451,7 +453,9 @@
         "_channels\\.tsv",
         "_ieeg\\.json",
         "_coordsystem\\.json",
-        "_photo\\.jpg"
+        "_photo\\.jpg",
+        "_photo\\.png",
+        "_photo\\.tif"
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
        "_photo\\.jpg"
```
Same as above, this change does not apply to "ieeg".

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 4]
```diff
@@ -481,6 +485,8 @@
         "_meg\\.json",
         "_coordsystem\\.json",
         "_photo\\.jpg",
+        "_photo\\.png",
+        "_photo\\.tif",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
```
Same as above, does not apply to "meg".

File: bids-validator/bids_validator/rules/session_level_rules.json [hunk 5]
```diff
@@ -77,6 +77,8 @@
         "_meg.json",
         "_coordsystem.json",
         "_photo.jpg",
+        "_photo.png",
+        "_photo.tif",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/session_level_rules.json]:
```suggestion
```
Doesn't apply to "meg".

File: bids-validator/bids_validator/rules/session_level_rules.json [hunk 6]
```diff
@@ -93,7 +95,9 @@
         "_electrodes.json",
         "_eeg.json",
         "_coordsystem.json",
-        "_photo.jpg"
+        "_photo.jpg",
+        "_photo.png",
+        "_photo.tif"
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/session_level_rules.json]:
```suggestion
        "_photo.jpg"
```
Doesn't apply to "eeg".

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 7]
```diff
@@ -526,7 +532,7 @@
   },
 
   "microscopy": {
-    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?microscopy[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))|(_photo\\.jpg))$",
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?microscopy[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))(@@@_photo_ext_@@@))$",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]: Here we are going to have to split into 2 possible regex expressions, because the filename template for `_photo` is not the same as for the other suffix, from comment here: https://github.com/bids-standard/bids-validator/pull/1359#discussion_r759659733

1. For all the "microscopy_type" (TEM, SEM, uCT, etc):
```
 "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?microscopy[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))$"
```
See reference here: https://bids-specification--881.org.readthedocs.build/en/881/04-modality-specific-files/10-microscopy.html#microscopy-imaging-data

2. For the `_photo` suffix:
```
"regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?microscopy[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(_photo)(@@@_photo_ext_@@@))$"
```

See reference here: https://bids-specification--881.org.readthedocs.build/en/881/04-modality-specific-files/10-microscopy.html#photos-of-the-samples-_photoextension

My syntax is probably not correct, let me know if you have questions.

File: bids-validator/bids_validator/rules/session_level_rules.json [hunk 8]
```diff
@@ -156,7 +160,9 @@
         "_electrodes.json",
         "_ieeg.json",
         "_coordsystem.json",
-        "_photo.jpg"
+        "_photo.jpg",
+        "_photo.png",
+        "_photo.tif"
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/session_level_rules.json]:
```suggestion
        "_photo.jpg"
```
Doesn't apply to "ieeg".

File: bids-validator/bids_validator/rules/top_level_rules.json [hunk 9]
```diff
@@ -97,6 +97,8 @@
         "_electrodes\\.tsv",
         "_electrodes\\.json",
         "_photo\\.jpg",
+        "_photo\\.png",
+        "_photo\\.tif",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/top_level_rules.json]:
```suggestion
```
Doesn't apply to "eeg".

File: bids-validator/bids_validator/rules/top_level_rules.json [hunk 10]
```diff
@@ -111,6 +113,8 @@
...
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/top_level_rules.json]:
```suggestion
```
Doesn't apply to "ieeg".

File: bids-validator/bids_validator/rules/top_level_rules.json [hunk 11]
```diff
@@ -123,6 +127,8 @@
         "_channels\\.tsv",
         "_channels\\.json",
         "_photo\\.jpg",
+        "_photo\\.png",
+        "_photo\\.tif",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/top_level_rules.json]:
```suggestion
```
Doesn't apply to "meg".

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 7 above)
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?micr[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))(@@@_photo_ext_@@@))$",
```

You may want to change `microscopy` to `micr` in this PR as well, as you prefer.

**Review by mariehbourget — COMMENTED** [review]
Thanks @konantian!
I left some clarification on a few points regarding the "photo" extensions.

You may want to use this PR to change "microscopy" to "micr" as well, as you prefer.

Also, I'm a bit confused about this:
image

It seems that the bep031 branch may not be up-to-date?
I'm not an expert here, but I don't understand how the same branch can be at different "state between the neuropoly fork and the buds-standard repo.
Maybe we are out-of-date with master?

**Comment by konantian** [comment]
> Thanks @konantian! I left some clarification on a few points regarding the "photo" extensions.
> 
> You may want to use this PR to change "microscopy" to "micr" as well, as you prefer.
> 
> Also, I'm a bit confused about this: image
> 
> It seems that the bep031 branch may not be up-to-date? I'm not an expert here, but I don't understand how the same branch can be at different "state between the neuropoly fork and the buds-standard repo. Maybe we are out-of-date with master?

The reason is after we forked our own repository from bids-standard, two repositories are exactly the same. However, with the update in the original branch, all the updates cannot be synchronized to our branch because they are two independent branches right now. If we want to keep our branch up-to-date, we have to manually synchronize two branches periodically.

File: bids-validator/validators/json/schemas/microscopy.json (hunk 1 above)
Comment (konantian) [reply bids-validator/validators/json/schemas/microscopy.json]: Done

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 3 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/file_level_rules.json]: Done

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 4 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/file_level_rules.json]: Done

File: bids-validator/bids_validator/rules/session_level_rules.json (hunk 8 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/session_level_rules.json]: Done

File: bids-validator/bids_validator/rules/top_level_rules.json (hunk 10 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/top_level_rules.json]: Done

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 2 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/file_level_rules.json]: Done

File: bids-validator/bids_validator/rules/session_level_rules.json (hunk 5 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/session_level_rules.json]: Done

File: bids-validator/bids_validator/rules/top_level_rules.json (hunk 9 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/top_level_rules.json]: Done

File: bids-validator/bids_validator/rules/top_level_rules.json (hunk 11 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/top_level_rules.json]: Done

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 7 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/file_level_rules.json]: Done

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 12]
```diff
@@ -526,7 +526,7 @@
   },
 
   "microscopy": {
-    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?microscopy[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))|(_photo\\.jpg))$",
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?micr[\\/\\\\](sub-[a-zA-Z0-9]+)(?:(_ses-[a-zA-Z0-9]+))?(?:_sample-[a-zA-Z0-9]+)(?:_acq-[a-zA-Z0-9]+)?(?:_stain-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_chunk-[a-zA-Z0-9]+)?(((@@@_microscopy_type_@@@)(@@@_microscopy_ext_@@@))|(@@@_photo_ext_@@@))$",
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json]: Is it possible to split the regex in 2 (1 for the `microscopy_ext` and 1 for the `photo_ext`)?

The reason is that microscopy files can have this form:
`sub-<label>[_ses-<label>]_sample-<label>[_acq-<label>][_stain-<label>][_run-<index>][_chunk-<index>]_<suffix>.<microscopy_ext>`
or this form:
`sub-<label>[_ses-<label>]_sample-<label>[_acq-<label>]_photo.<photo_ext>`
where `[]` are for optional `key-value` pairs.

So all the `key-value` pairs are not valid for the `_photo` suffix. The validator should check if the filename correspond to one template or the other.

Note: this does not apply to the `json` extension which is OK as is.

**Review by mariehbourget — COMMENTED** [review]
Thanks @konantian!
There is a minor adjustment to do to the regex for `photo`.

After this PR, would you be able to take care of synchronizing our bep031 branch with master?

File: bids-validator/bids_validator/rules/file_level_rules.json (hunk 12 above)
Comment (konantian) [reply bids-validator/bids_validator/rules/file_level_rules.json]: @mariehbourget It's definitely possible to split the regex in 2 and I have done this already. To examine the new regex, I renamed the data file under micr_SEM from `sub-01_ses-01_sample-A_SEM.png` to `sub-01_ses-01_sample-A_photo.png` and remained the JSON file `sub-01_ses-01_sample-A_SEM.json` unchanged. With this change, the png file could be recognized correctly by the validator as a microscopy file. However,  an error raised 

`1: [ERR] A json sidecar file was found without a corresponding data file (code: 90 - SIDECAR_WITHOUT_DATAFILE)
                ./sub-01/ses-01/micr/sub-01_ses-01_sample-A_SEM.json
` when I run the validator with `micr_SEM` dataset.

From my perspective, both data and json files followed the naming rules. In this case, this error should not be raised right? Please correct me if I am mistaken.

Comment (mariehbourget) [reply bids-validator/bids_validator/rules/file_level_rules.json]: Hi @konantian!

The way I understand the regex now, it works perfectly for all the suffixes that are not `_photo` (for all suffixes listed in `microscopy_type`, ex: `_SEM`).
But with the suffix `_photo`, a name like `sub-01_ses-01_sample-01_acq-01_stain-01_run-01_chunk-01_photo.jpg` would also pass the validation, which is not correct because `stain`, `run` and `chunk` are not allowed with the photo suffix.

For the error in JSON, it is normal because elsewhere in the validator, they check if each json file has a corresponding image file, which was not the case with your example. `_photo` files don't have json file, so don't worry about that.

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 13]
```diff
@@ -555,8 +555,18 @@
       ]
     }
   },
+  "microscopy_photo": {
```
Comment (mariehbourget) [inline bids-validator/bids_validator/rules/file_level_rules.json:558]: Looks good, thank you!

File: bids-validator/utils/type.js [hunk 14]
```diff
@@ -390,10 +394,10 @@ export default {
   },
 
   checkType(obj, typeString) {
-    if (typeString == 'number') {
+    if (typeString === 'number') {
```
Comment (mariehbourget) [inline bids-validator/utils/type.js]: Is this change related to Microscopy specifically?
If not, I would recommend not making changes to the already existing code of the validator, to not impact other modalities' validation.

File: bids-validator/utils/type.js [hunk 15]
```diff
@@ -390,10 +394,10 @@ export default {
...
       return !isNaN(parseFloat(obj)) && isFinite(obj)
     } else {
-      return typeof obj == typeString
+      return typeof obj === typeString
```
Comment (mariehbourget) [inline bids-validator/utils/type.js]: Same as above

File: bids-validator/validators/json/schemas/microscopy.json [hunk 16]
```diff
@@ -62,9 +62,9 @@
...
     "InstitutionAddress",
     "InstitutionalDepartmentName",
-
     "BodyPart",
-    "Environment",
+    "BodyPartDetails",
+    "SampleEnvironment",
     "SampleStaining",
     "SamplePrimaryAntibody",
     "SampleSecondaryAntibody"
```
Comment (mariehbourget) [inline bids-validator/validators/json/schemas/microscopy.json:70]: Just a question, why was `ChunkTransformationMatrix` removed from the list of recommended?
It is recommended if `chunk-<index>` is present in filenames (unless I made a mistake somewhere).
Just let me know, thanks!

Comment (konantian) [reply bids-validator/validators/json/schemas/microscopy.json:70]: `ChunkTransformationMatrix` is removed from here because there's no way to check if `chunk-<index>` is present in the filenames in the microscopy schema. My solution is to check this in fullTest.js and if `ChunkTransformationMatrix` does not exist in the JSON file, there will be a warning raised.

Comment (mariehbourget) [reply bids-validator/validators/json/schemas/microscopy.json:70]: Perfect, thanks for the explanation!

**Review by mariehbourget — APPROVED** [review]
